modules/common/react/lib/utils/components.ts

```diff
@@ -0,0 +1,184 @@
+  displayName?: string;
+  /** The component function. The function looks like:
+   * @example
+   * Component: ({children}, ref, Element) {
+   *   // `Element` is what's passed to the `as` of your component. It will be 'div' or even a another Component!
+   *   return (
+   *     <Element ref={ref}>{children}</Element>
+   *   )
+   * }
+   */
+  Component: RefForwardingComponent<T, P>;
+  /**
+   * Used in container components
+   */
+  subComponents?: SubComponents;
+}) => {
+  const ReturnedComponent = (React.forwardRef<T, P & {as?: React.ElementType}>(
+    ({as: asOverride, ...props}, ref) => {
+      return Component(
+        props as P,
+        //@ts-ignore Problems with type constraints of T and T extends ElementTagName ? ... This doesn't cause a problem, not sure it is worth fixing
+        ref,
+        asOverride || as
+      );
+    }
+  ) as any) as Component<T, P> & SubComponents;
+
+  Object.keys(subComponents || {}).forEach((key: any) => {
+    // @ts-ignore Not worth typing correctly
+    ReturnedComponent[key] = subComponents[key];
+  });
+  ReturnedComponent.displayName = displayName;
+
+  return ReturnedComponent;
+};
+
+function setRef<T>(ref: React.Ref<T>, value: T): void {
+  if (ref) {
+    if (typeof ref === 'function') {
+      ref(value);
+    } else {
+      // @ts-ignore Refs are readonly, but we can technically write to it without issue
+      ref.current = value;
+    }
+  }
+}
+
+/**
+ * This function will create a new forked ref out of two input Refs. This is useful for components
+ * that use `React.forwardRef`, but also need internal access to a Ref.
+ *
+ * This function is inspired by https://www.npmjs.com/package/@rooks/use-fork-ref
+ *
+ * @example
+ * React.forwardRef((props, ref) => {
+ *   // Returns a RefObject with a `current` property
+ *   const myRef = React.useRef(ref)
+ *
+ *   // Returns a forked Ref function to pass to an element.
+ *   // This forked ref will update both `myRef` and `ref` when React updates the element ref
+ *   const elementRef = useForkRef(ref, myRef)
+ *
+ *   useEffect(() => {
+ *     console.log(myRef.current) // `current` is the DOM instance
+ *     // `ref` might be null since it depends on if someone passed a `ref` to your component
+ *     // `elementRef` is a function and we cannot get a current value out of it
+ *   })
+ *
+ *   return <div ref={elementRef}/>
+ * })
+ */
+export function useForkRef<T>(ref1: React.Ref<T>, ref2: React.Ref<T>): (instance: T) => void {
+  return (value: T) => {
+    setRef(ref1, value);
+    setRef(ref2, value);
+  };
+}
+
+/**
+ * This function is the same as calling `React.useRef(null)` with the added benefit of extracting
+ * the type from the ref passed to it. This means you don't have to provide a generic to the
+ * `useRef` function
+ * @param ref The React ref passed from the `createComponent` factory function
+ */
+export function useComponentRef<T>(ref: React.Ref<T>) {
+  return React.useRef<T>(null);
+}
```

Comment:
I've been using these together any time I need a local ref for something:

```tsx
  Component: ({name = '', model, children, ...elemProps}: TabProps, ref, Element) => {
    const localRef = useComponentRef(ref);
    const elementRef = useForkRef(ref, localRef);

    // do something with localRef
    return <Element ref={elementRef} ... />
```

Since it is so common to use them together and it feels odd to juggle so many refs, why not a function that combines them?

```tsx
  Component: ({name = '', model, children, ...elemProps}: TabProps, ref, Element) => {
    const { localRef, elementRef } = useLocalRef(ref);

    // do something with localRef
    return <Element ref={elementRef} ... />
```